frame/balances/src/lib.rs

```diff
@@ -602,1601 +466,670 @@ pub mod pallet {
 			);
 
 			for &(ref who, free) in self.balances.iter() {
+				frame_system::Pallet::<T>::inc_providers(who);
 				assert!(T::AccountStore::insert(who, AccountData { free, ..Default::default() })
 					.is_ok());
 			}
 		}
 	}
-}
-
-#[cfg(feature = "std")]
-impl<T: Config<I>, I: 'static> GenesisConfig<T, I> {
-	/// Direct implementation of `GenesisBuild::build_storage`.
-	///
-	/// Kept in order not to break dependency.
-	pub fn build_storage(&self) -> Result<sp_runtime::Storage, String> {
-		<Self as GenesisBuild<T, I>>::build_storage(self)
-	}
-
-	/// Direct implementation of `GenesisBuild::assimilate_storage`.
-	///
-	/// Kept in order not to break dependency.
-	pub fn assimilate_storage(&self, storage: &mut sp_runtime::Storage) -> Result<(), String> {
-		<Self as GenesisBuild<T, I>>::assimilate_storage(self, storage)
-	}
-}
 
-/// Simplified reasons for withdrawing balance.
-#[derive(Encode, Decode, Clone, Copy, PartialEq, Eq, RuntimeDebug, MaxEncodedLen, TypeInfo)]
-pub enum Reasons {
-	/// Paying system transaction fees.
-	Fee = 0,
-	/// Any reason other than paying system transaction fees.
-	Misc = 1,
-	/// Any reason at all.
-	All = 2,
-}
+	#[cfg(feature = "std")]
+	impl<T: Config<I>, I: 'static> GenesisConfig<T, I> {
+		/// Direct implementation of `GenesisBuild::build_storage`.
+		///
+		/// Kept in order not to break dependency.
+		pub fn build_storage(&self) -> Result<sp_runtime::Storage, String> {
+			<Self as GenesisBuild<T, I>>::build_storage(self)
+		}
 
-impl From<WithdrawReasons> for Reasons {
-	fn from(r: WithdrawReasons) -> Reasons {
-		if r == WithdrawReasons::TRANSACTION_PAYMENT {
-			Reasons::Fee
-		} else if r.contains(WithdrawReasons::TRANSACTION_PAYMENT) {
-			Reasons::All
-		} else {
-			Reasons::Misc
+		/// Direct implementation of `GenesisBuild::assimilate_storage`.
+		///
+		/// Kept in order not to break dependency.
+		pub fn assimilate_storage(&self, storage: &mut sp_runtime::Storage) -> Result<(), String> {
+			<Self as GenesisBuild<T, I>>::assimilate_storage(self, storage)
 		}
 	}
-}
 
-impl BitOr for Reasons {
-	type Output = Reasons;
-	fn bitor(self, other: Reasons) -> Reasons {
-		if self == other {
-			return self
+	#[pallet::call]
+	impl<T: Config<I>, I: 'static> Pallet<T, I> {
+		/// Transfer some liquid free balance to another account.
+		///
+		/// `transfer_allow_death` will set the `FreeBalance` of the sender and receiver.
+		/// If the sender's account is below the existential deposit as a result
+		/// of the transfer, the account will be reaped.
+		///
+		/// The dispatch origin for this call must be `Signed` by the transactor.
+		///
+		/// # <weight>
```

Comment:
FYI I am pushing for getting rid of `<weight>` comments, ie. https://github.com/paritytech/substrate/pull/13350. You can skip them now or just put them under `## Complexity` if it is useful for the developers to know. If it is only for us, I'd put it under `//` as *implementation note*.